core/domain/caching_services_test.py

```diff
@@ -0,0 +1,368 @@
+            deserialize(serialize(default_exploration)).to_dict())
+
+    def test_invalid_namespace_raises_error(self):
+        invalid_namespace = 'invalid'
+        key_value_mapping = {'a': '1', 'b': '2', 'c': '3'}
+
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Invalid namespace: %s.' % invalid_namespace):
+            caching_services.set_multi(
+                invalid_namespace, None,
+                key_value_mapping)
+
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Invalid namespace: %s.' % invalid_namespace):
+            caching_services.get_multi(
+                invalid_namespace, None,
+                ['a', 'b', 'c'])
+
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Invalid namespace: %s.' % invalid_namespace):
+            caching_services.delete_multi(
+                invalid_namespace, None, ['a', 'b', 'c'])
+
+        invalid_sub_namespace = 'sub:namespace'
+        with self.assertRaisesRegexp(
+            ValueError,
+            'Sub-namespace %s cannot contain \':\'.' % invalid_sub_namespace):
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_DEFAULT,
+                invalid_sub_namespace, ['a', 'b', 'c'])
+
+    def test_get_multi_correctly_retrieves_cache_elements(self):
+        """Testing that querying the cache for elements where either all of the
+        ids exist or don't exist in the cache returns reasonable output.
+        """
+        key_value_mapping = {'a': '1', 'b': '2', 'c': '3'}
+
+        exploration_id = 'id'
+
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_DEFAULT, None,
+                ['a', 'b', 'c']), {})
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_EXPLORATION,
+                python_utils.convert_to_bytes(0),
+                [exploration_id]),
+            {})
+
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, key_value_mapping)
+        default_exploration = (
+            exp_domain.Exploration.create_default_exploration(
+                'exp_id_1', title='A title', category='A category'))
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_EXPLORATION,
+            python_utils.convert_to_bytes(0),
+            {
+                exploration_id: default_exploration
+            })
+        keys_exist_result = caching_services.get_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, ['a', 'b', 'c'])
+        self.assertEqual(keys_exist_result, key_value_mapping)
+        keys_non_existent_result = caching_services.get_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, ['d', 'e'])
+        self.assertEqual(keys_non_existent_result, {})
+
+        exploration_dict = caching_services.get_multi(
```

Comment:
This name is misleading since the values of this dict are themselves "explorations" which individually become "exploration dicts". Perhaps exp_ids_to_explorations?